rust/src/dns/detect.rs

```diff
@@ -103,6 +121,363 @@ pub extern "C" fn SCDnsDetectRrtypeMatch(
+    de: *mut c_void, transforms: *const c_void, flow: *const c_void, flow_flags: u8,
+    tx: *const c_void, list_id: c_int, local_id: u32,
+) -> *mut c_void {
+    return DetectHelperGetMultiData(
+        de,
+        transforms,
+        flow,
+        flow_flags,
+        tx,
+        list_id,
+        local_id,
+        dns_tx_get_query_name,
+    );
+}
+
+unsafe extern "C" fn dns_query_get_data_wrapper(
+    de: *mut c_void, transforms: *const c_void, flow: *const c_void, flow_flags: u8,
+    tx: *const c_void, list_id: c_int, local_id: u32,
+) -> *mut c_void {
+    return DetectHelperGetMultiData(
+        de,
+        transforms,
+        flow,
+        flow_flags,
+        tx,
+        list_id,
+        local_id,
+        dns_tx_get_query,
+    );
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCDetectDNSRegister() {
+    let kw = SCSigTableElmt {
+        name: b"dns.answer.name\0".as_ptr() as *const libc::c_char,
+        desc: b"DNS answer name sticky buffer\0".as_ptr() as *const libc::c_char,
+        url: b"/rules/dns-keywords.html#dns-answer-name\0".as_ptr() as *const libc::c_char,
+        Setup: dns_detect_answer_name_setup,
+        flags: SIGMATCH_NOOPT | SIGMATCH_INFO_STICKY_BUFFER,
+        AppLayerTxMatch: None,
+        Free: None,
+    };
+    let _g_dns_answer_name_kw_id = DetectHelperKeywordRegister(&kw);
+    G_DNS_ANSWER_NAME_BUFFER_ID = DetectHelperMultiBufferMpmRegister(
+        b"dns.answer.name\0".as_ptr() as *const libc::c_char,
+        b"dns answer name\0".as_ptr() as *const libc::c_char,
+        ALPROTO_DNS,
+        true,
+        /* Register also in the TO_SERVER direction, even though this is not
+        normal, it could be provided as part of a request. */
+        true,
+        dns_answer_name_get_data_wrapper,
+    );
+    let kw = SCSigTableElmt {
+        name: b"dns.opcode\0".as_ptr() as *const libc::c_char,
+        desc: b"Match the DNS header opcode flag.\0".as_ptr() as *const libc::c_char,
+        url: b"rules/dns-keywords.html#dns-opcode\0".as_ptr() as *const libc::c_char,
+        AppLayerTxMatch: Some(dns_opcode_match),
+        Setup: dns_opcode_setup,
+        Free: Some(dns_opcode_free),
+        flags: 0,
+    };
+    G_DNS_OPCODE_KW_ID = DetectHelperKeywordRegister(&kw);
+    G_DNS_OPCODE_BUFFER_ID = DetectHelperBufferRegister(
+        b"dns.opcode\0".as_ptr() as *const libc::c_char,
+        ALPROTO_DNS,
+        true,
+        true,
+    );
+    let kw = SCSigTableElmt {
+        name: b"dns.query.name\0".as_ptr() as *const libc::c_char,
+        desc: b"DNS query name sticky buffer\0".as_ptr() as *const libc::c_char,
+        url: b"/rules/dns-keywords.html#dns-query-name\0".as_ptr() as *const libc::c_char,
+        Setup: dns_detect_query_name_setup,
+        flags: SIGMATCH_NOOPT | SIGMATCH_INFO_STICKY_BUFFER,
+        AppLayerTxMatch: None,
+        Free: None,
+    };
+    let _g_dns_query_name_kw_id = DetectHelperKeywordRegister(&kw);
```

Comment:
Have you looked at making registering keywords more of a Rust experience? Rather than this C like code from Rust?  By that I mean registration functions where you can use Rust strings, and hide the as_ptr() and stuff away.

Reply:
Good idea, will do

Reply:
while at it, lets give `SCSigTableElmt` a better name. It's currently a app-layer specific "lite" version of the full C SigTableElmt, which has got me confused several times now.

Reply:
Trying something